File: MB-Secretaria/observability_py/drift_detector.py
from datetime import UTC, datetime

from .types import DriftSignal, ObservabilityEvent, ObservabilityLayer, ObservabilitySeverity


def _signal(layer: ObservabilityLayer, signal_type: str, severity: ObservabilitySeverity, description: str, events: list[ObservabilityEvent]) -> DriftSignal:
    return DriftSignal(layer, signal_type, severity, description, [event.trace_id for event in events[:10]], datetime.now(UTC).isoformat())


def detect_drift_signals(events: list[ObservabilityEvent]) -> list[DriftSignal]:
    signals: list[DriftSignal] = []
    fallbacks = [event for event in events if "fallback" in event.event_type]
    low_confidence = [event for event in events if event.layer == "confidence" and float(event.payload_summary.get("confidence_score", 1)) < 0.5]
    degraded = [event for event in events if event.layer == "provider" and "degraded" in event.event_type]
    review_escalations = [event for event in events if event.layer == "review" and "escalation" in event.event_type]
    multimodal = [event for event in events if "multimodal_conflict" in event.event_type]
    if len(fallbacks) >= 3:
        signals.append(_signal("system", "fallback_spike", "warn", "Fallback events are increasing.", fallbacks))
    if len(low_confidence) >= 3:
        signals.append(_signal("confidence", "low_confidence_spike", "warn", "Low confidence events are increasing.", low_confidence))
    if len(degraded) >= 2:
        signals.append(_signal("provider", "provider_degradation", "error", "Provider degraded events detected.", degraded))
    if len(review_escalations) >= 2:
        signals.append(_signal("review", "review_escalation_spike", "warn", "Human review escalation pressure increased.", review_escalations))
    if multimodal:
        signals.append(_signal("imaging", "multimodal_conflict", "warn", "Multimodal conflict signal detected.", multimodal))
    return signals

File: MB-Secretaria/observability/drift-detector.ts
import type { DriftSignal, ObservabilityEvent } from "./types";

export function detectDriftSignals(events: ObservabilityEvent[]): DriftSignal[] {
  const now = new Date().toISOString();
  const signals: DriftSignal[] = [];
  const fallbackEvents = events.filter((event) => event.event_type.includes("fallback"));
  const lowConfidence = events.filter((event) => event.layer === "confidence" && Number(event.payload_summary.confidence_score) < 0.5);
  const degradedProviders = events.filter((event) => event.layer === "provider" && event.event_type.includes("degraded"));
  const reviewEscalations = events.filter((event) => event.layer === "review" && event.event_type.includes("escalation"));
  const multimodalConflicts = events.filter((event) => event.event_type.includes("multimodal_conflict"));

  if (fallbackEvents.length >= 3) signals.push(buildSignal("system", "fallback_spike", "warn", "Fallback events are increasing.", fallbackEvents, now));
  if (lowConfidence.length >= 3) signals.push(buildSignal("confidence", "low_confidence_spike", "warn", "Low confidence events are increasing.", lowConfidence, now));
  if (degradedProviders.length >= 2) signals.push(buildSignal("provider", "provider_degradation", "error", "Provider degraded events detected.", degradedProviders, now));
  if (reviewEscalations.length >= 2) signals.push(buildSignal("review", "review_escalation_spike", "warn", "Human review escalation pressure increased.", reviewEscalations, now));
  if (multimodalConflicts.length >= 1) signals.push(buildSignal("imaging", "multimodal_conflict", "warn", "Multimodal conflict signal detected.", multimodalConflicts, now));
  return signals;
}

function buildSignal(layer: DriftSignal["layer"], signal_type: string, severity: DriftSignal["severity"], description: string, events: ObservabilityEvent[], now: string): DriftSignal {
  return { layer, signal_type, severity, description, trace_refs: events.map((event) => event.trace_id).slice(0, 10), detected_at: now };
}
